### PR TITLE
Update profiling records without reseting catalog

### PR DESCRIPTION
**Pull request by outtanames**
Variety of updates/fixes to the profiler to support generating profiling catalogs across local/cloud:
* profiling runs now load and update the existing catalog rather than regenerating the entire thing, so we can do incremental model profiling runs for a specific subset of models
* Various fixes/checks when reading/writing to/from the profiling catalog, edge cases were breaking e.g. gpu utilization
* Refactor profiler to make different overloads of `prof` less ambiguous between the Profiler itself, context managers, actual profiling results etc.
* `--catalog-path` flag so we can write profiling results to a filemount in skypilot rather than needing to track down the catalog inside the nos cache each time.

Screenshot 2024-01-22 at 3 34 28 PM

## Checks

- [ ] `make lint`: I've run `make lint` to lint the changes in this PR.
- [ ] `make test`: I've made sure the tests (`make test-cpu` or `make test`) are passing.
- Additional tests:
   - [ ] Benchmark tests (when contributing new models)
   - [ ] GPU/HW tests

### CODE REVIEW

**Comment by spillai** [comment]
Can you add tests for these?

File: nos/common/profiler.py [hunk 1]
```diff
@@ -228,6 +228,22 @@ def as_df(self) -> pd.DataFrame:
         )
         return df
 
+    def from_df(self, df: pd.DataFrame) -> None:
```
Comment (spillai) [inline nos/common/profiler.py:231]: This needs to be a `classmethod` where you `return cls(records)`

File: nos/common/profiler.py [hunk 2]
```diff
@@ -228,6 +228,22 @@ def as_df(self) -> pd.DataFrame:
...
+        self.records = []
+        for _idx, row in df.iterrows():
+            namespace = row["namespace"]
+            kwargs = row.drop("namespace").to_dict()
+            record = profiler_record(namespace, **kwargs)
+            for key, value in row.items():
+                record.update(key, value)
+            self.records.append(record)
+
+    def from_json_path(self, filename: Union[Path, str]) -> None:
```
Comment (spillai) [inline nos/common/profiler.py:242]: Same comment as above, needs to be a `classmethod` that returns `return cls.from_df(df)`

File: nos/common/profiler.py (hunk 1 above)
Comment (spillai) [reply nos/common/profiler.py:231]:
```python
@classmethod
def from_df(cls, df: pd.DataFrame) -> "Profiler":
    ...
    return cls(records)
```

File: nos/common/profiler.py [hunk 3]
```diff
@@ -295,6 +311,15 @@ def __repr__(self) -> str:
...
         """Setup the device and profiler."""
 
+        # Check if we have a previous catalog and load it
+        from nos.constants import NOS_PROFILE_CATALOG_PATH
+
+        if NOS_PROFILE_CATALOG_PATH.exists():
+            self.prof = Profiler()
+            self.prof.from_json_path(NOS_PROFILE_CATALOG_PATH)
+        else:
+            logger.debug("No prof catalog found")
```
Comment (spillai) [inline nos/common/profiler.py:321]: Use `f"Profile catalog not found (filename={NOS_PROFILE_CATALOG_PATH})."`

File: nos/common/profiler.py [hunk 4]
```diff
@@ -403,9 +428,12 @@ def run(self) -> None:
         failed = {}
         st_t = time.time()
 
-        print()
         print(f"[white]{self}[/white]")
-        with Profiler() as self.prof, torch.inference_mode():
+        from nos.constants import NOS_PROFILE_CATALOG_PATH
+
+        self.prof = Profiler()
```
Comment (spillai) [inline nos/common/profiler.py:434]: self.prof = Profiler.from_json_path(NOS_PROFILE_CATALOG_PATH)

**Comment by outtanames** [comment]
> Can you add tests for these?

see additions, we now check for gpu utilization in the catalog for `clip-vit-patch32` (along with a successful profiling run overall).

**Pull request merged**